### PR TITLE
InputBase: Simplify focus styles

### PR DESCRIPTION
**Pull request by mirka**
## What?

Simplifies how we coordinate focus styles in the `InputBase` primitive component so it can be done "automatically" with just CSS, without relying on React props/state.

No visual/behavioral changes are intended.

## Why?

InputBase is an internal component used to style the standard borders for an input, as well as handle the layout for prefix/suffix elements.

I'd like to reuse InputBase for styling CustomSelectControlV2, like we currently do with v1. This is a good opportunity to simplify this focus handling in there, lest we add even more unnecessary state management.

## How?

Use `:focus-within` selectors to detect inner focus declaratively, rather than keeping track of it imperatively through focus/blur listeners.

## Testing Instructions

There shouldn't have been any visual changes in focus behavior.

### Affected components worth testing

- InputControl — Also test with the `disabled` prop enabled.
- CustomSelectControl (v1) — Given the non-trivial changes in this one, it should be tested in the editor as well, for example the Appearance dropdown in the block inspector Typography tools, or the Position dropdown in the block inspector of a Row block.
- UnitControl — When focus moves to the unit selector dropdown, the focus indicator should also move there.
- SearchControl — Uses the borderless variant of InputBase.
- ColorPicker — The color format dropdown (Hex/RGB/HSL) uses the borderless variant of InputBase.

### Affected but low chance of regressions

- NumberControl
- SelectControl
- TreeSelect

#### Affected through style overrides of relevant elements

- The time input fields in TimePicker
- The color name input field when in Edit mode of PaletteEdit

### Testing Instructions for Keyboard

Test <kbd>Tab</kbd> and arrow keys to navigate elements, especially UnitControl and CustomSelectControl (v1).

### CODE REVIEW

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (mirka) [inline packages/components/src/input-control/styles/input-control-styles.tsx:36]: No changes, just moved higher up the file so we can refer to them earlier.

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (mirka) [inline packages/components/src/input-control/styles/input-control-styles.tsx:290]: The important priority logic in here is for `border-color`, where `isFocused` > `isBorderless` > `disabled` (accounting for the fact that a disabled control will never be focused). This priority logic still holds true in the refactored styles.

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (mirka) [inline packages/components/src/input-control/styles/input-control-styles.tsx:86]: Not sure why this `z-index` is necessary, but will leave it in for the scope of this refactor PR.

File: packages/components/src/input-control/input-base.tsx
Comment (mirka) [inline packages/components/src/input-control/input-base.tsx:113]: Removed because the `labelPosition` prop was not being used.

File: packages/components/src/custom-select-control/index.js
Comment (mirka) [inline packages/components/src/custom-select-control/index.js:228]: Rejiggers the DOM structure a bit so the focus styles can be managed with CSS only.

**Comment by DaniGuardiola** [comment]
I'm in the middle of reviewing this, please leave it to me (in case someone was thinking of taking over) :)

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (tyxla) [inline packages/components/src/input-control/styles/input-control-styles.tsx:64]: Should this be using `CONFIG.borderWidth`?

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (tyxla) [inline packages/components/src/input-control/styles/input-control-styles.tsx:93]: We have this contrast mode enhancement at least in a few places already - should we make it reusable?

**Review by tyxla — COMMENTED** [review]
Leaving a couple of drive-by comments as I was glancing through it.

I still haven't done a thorough review or testing as of yet. 

> I'm in the middle of reviewing this, please leave it to me (in case someone was thinking of taking over) :)

Sounds great, go for it @DaniGuardiola 🙌

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (DaniGuardiola) [reply packages/components/src/input-control/styles/input-control-styles.tsx:93]: Since we're migrating away from emotion I'd suggest holding off on creating any abstractions, at least until a unified styling solution is decided.

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (DaniGuardiola) [inline packages/components/src/input-control/styles/input-control-styles.tsx:85]: I'm unfamiliar with the browser versions we're targeting for support but beware that `:has` is recent enough to still not be supported for some potential users: https://caniuse.com/css-has

**Review by DaniGuardiola — APPROVED** [review]

**Review by DaniGuardiola — APPROVED** [review]
Makes sense to me. My only concern is about `:has` support, but I'm all for relying on it.

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (DaniGuardiola) [reply packages/components/src/input-control/styles/input-control-styles.tsx:85]: Related thought but, are we only targeting the input for focus here? Because, in that case, a simpler selector like `:has(input:focus)` might work and be easier both on maintainers and the browser itself.

**Comment by tyxla** [comment]
> Makes sense to me. My only concern is about `:has` support, but I'm all for relying on it.

`:has` should be fine to use nowadays. According to the WP [browser support](https://make.wordpress.org/core/handbook/best-practices/browser-support/) it is supported on all browsers, except for older Safari on iOS which still in total has the borderline 1% support.

File: packages/components/src/dimension-control/test/__snapshots__/index.test.js.snap
Comment (tyxla) [inline packages/components/src/dimension-control/test/__snapshots__/index.test.js.snap:50]: Doesn't look like we particularly care about those styles. Would be nice if we alter the snapshot test to something more specific. Not a blocker for this PR.

File: packages/components/CHANGELOG.md
Comment (tyxla) [inline packages/components/CHANGELOG.md]: Should we list them all? May be easier if someone is looking for the changelog in the case of a regression.

**Review by tyxla — APPROVED** [review]
All affected components tested well for me in storybook. ✅ 
I did some testing in the post editor and site editor as well - no surprises. 

Nice work 🚀

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (mirka) [reply packages/components/src/input-control/styles/input-control-styles.tsx:64]: I don't see the point of having that be a global variable to be honest, it may be on the chopping block for #43997.

File: packages/components/CHANGELOG.md
Comment (mirka) [reply packages/components/CHANGELOG.md]: It's impractical to list all of them because so many compound components include some form of these inputs, but I added a few more to cover all the relatively primitive ones. I think that should be sufficient because those are the ones more likely to involve consumer overrides.

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (mirka) [reply packages/components/src/input-control/styles/input-control-styles.tsx:85]: > in that case, a simpler selector like `:has(input:focus)` might work

It's tempting, but technically the prefix/suffix slots can also contain `input` elements, so no 🥲 Also the primary focusable element in the InputBase isn't always an `input`, as is the case with SelectControl.

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (mirka) [reply packages/components/src/input-control/styles/input-control-styles.tsx:93]: Good that it's easy to grep, since they always have a code comment 😂

File: packages/components/src/dimension-control/test/__snapshots__/index.test.js.snap
Comment (mirka) [reply packages/components/src/dimension-control/test/__snapshots__/index.test.js.snap:50]: Agreed. I'm even questioning the validity of DimensionControl as a standalone component, but that's another story 🙃

**Comment by mirka** [comment]
Thanks for the reviews and testing! 🙏

File: packages/components/src/input-control/styles/input-control-styles.tsx
Comment (t-hamano) [reply packages/components/src/input-control/styles/input-control-styles.tsx:86]: I was researching the cause of #61931 and came across this PR.

This `z-index` seems to cause the slider dots below to be brought to the front when I open the Appearance dropdown. Removing this `z-index` seems to solve the problem, is there any better approach? [`z-index:3` is applied](https://github.com/WordPress/gutenberg/blob/62b0efd7af66c18d5717028e75cf8ff0c441c196/packages/block-editor/src/components/spacing-sizes-control/style.scss#L38) to this slider dots.

| z-index enabled | z-index disabled |
|--------|--------|
| image | image |

Comment (DaniGuardiola) [reply packages/components/src/input-control/styles/input-control-styles.tsx:86]: I haven't looked too deep into this, but I'd suggest simply making sure that the dots are the last elements in the tree so that they naturally take precedence in the z layer. But maybe I am missing some important information. @t-hamano please let me know if you find the root issue and a clean solution, I'd be happy to help otherwise. I want to make sure the fix doesn't create a problem down the road, as these z-index problems often do, unfortunately. Thanks for looking into this!

Comment (mirka) [reply packages/components/src/input-control/styles/input-control-styles.tsx:86]: Thanks for investigating, @t-hamano. Let's just remove the z-index like you suggest. I looked into the [blame](https://github.com/WordPress/gutenberg/blame/4964bf052c89add640657cf438405c051854a11b/packages/components/src/input-control/styles/input-control-styles.tsx#L34) for the original `z-index` declaration (#22329), and I think we can kind of deduce that it was for the "Floating label UI interaction", which is a feature we don't have in the codebase anymore.

**Pull request merged**
